linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_system_auth/oval/shared.xml

```diff
@@ -1,41 +1,148 @@
 <def-group>
-  <definition class="compliance" id="accounts_password_pam_pwhistory_remember_system_auth" version="2">
+  <definition class="compliance" id="{{{ rule_id }}}" version="2">
     {{{ oval_metadata("The passwords to remember should be set correctly.") }}}
-    <criteria comment="remember parameter of pam_pwhistory.so is set correctly" operator="AND">
-      <criterion comment="remember parameter of pam_pwhistory.so is set correctly" test_ref="test_accounts_password_pam_pwhistory_remember_system_auth" />
+    <criteria operator="AND" comment="Check if pam_pwhistory.so is properly configured">
+      <!--
+        pam_pwhistory.so parameters can be defined directly in pam files or, in newer versions,
+        in /etc/security/pwhistory.conf. The last is the recommended option when available. Also,
+        is the option used by auselect tool. However, regardless the approach, a minimal
+        declaration is common in pam files. -->
+      <criterion test_ref="test_accounts_password_pam_pwhistory_remember_system_auth"
+        comment="pam_pwhistory.so is properly defined in password section of system-auth"/>
+
+      <!--
+        pam_pwhistory.so parameters should be defined in /etc/security/pwhistory.conf whenever
+        possible. But due to backwards compatibility, they are also allowed in pam files directly.
+        In case they are defined in both places, pam files have precedence and this may confuse
+        the assessment. The following tests ensure only one option is used. Note that if
+        pwhistory.conf is available, authselect tool will preserve PAM files from parameters
+        definitions in favor of /etc/security/pwhistory.conf. -->
+      <criteria operator="OR"
```

Comment:
I think you can do some code deduplication here. If I understand it correctly, following criteria use two pairs of tests:
either the value is present in system-auth and not in pwhistory.conf
or the value is not present in system-auth and is present in pwhistory.conf
The <criterion> tag of OVAL offers the "negate" argument, which allows to specify the test in the same way as usual but negate its result. I think you can work here only with two test definitions instead of four.
You can see example here:
https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/bootloader-zipl/zipl_systemd_debug-shell_argument_absent/oval/shared.xml

Reply:
Hi Vojtech. This could be done. However, I preferred to make it more verbose. This OVAL logic is relatively complex comparing to most of the other rules, so using this approach it is easier to read and understand for a wider audience. I any case, I am also taking advantage of existing objects in these tests. Basically the tests are explicit to make them more readable, but the objects are shared.

Reply:
Maybe this could be considered in a separate PR indented to reduce code without increasing complexity. But I would prefer to not change this at the moment since any small change consumes a considerable amount of time testing the impact. I would prefer to think in code deduplication for this case only after the current logic is stable. Are you fine with that?

Reply:
OK I agree. The duplicated code is really small here, just two extra tests which both reuse already defined object.